Add reCAPTCHA v3 to user step on load, and submit token with n…

andrewserong opened the pull request:
Adds reCAPTCHA v3 to the user step in signup.

More reading: https://developers.google.com/recaptcha/docs/v3

When the user step mounts, it calls a function `initGoogleRecaptcha` that loads the reCAPTCHA script and when the user submits the form it calls `recordGoogleRecaptchaAction` to record and action that we will verify with the backend.

Firing the action when the button is clicked slows down form submission. Rationale is Google can only verify an action that's occured in the last 2 minutes, and the user could sit on the user step for a long time. So the action is recorded right as the user is ready to leave.

#### Changes proposed in this Pull Request

* Load recaptcha when user step mounts
* Send `calypso/signup/pageLoad` action when step mounts
* Send `calypso/signup/formSubmit` action when submitting the user step
* Sends the token from the above action to the new user API as the `g-recaptcha-response` param
* Recaptcha UI removed when leaving the user step.
* Doesn't do any of this when in the test environment because the recaptcha key is excluded from `test.json`
* Place A/B test behaviour behind a 50/50 A/B test `userStepRecaptcha`

We never verify the page load action with the backend so technically not needed. Maybe it helps train the AI? 🤷‍♂ 

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply D34725-code
* Go to http://calypso.localhost:3000/start
* From the bug icon in the bottom right, select the `show` variant of the `userStepRecaptcha` A/B test
* You should see the reCAPTCHA notice in the bottom right
* Open up the Network tab in DevTools and create a new user account
* Take a look at the network request for the API call to create a new user `/users/new` and you should see the `g-recaptcha-response` prop with the reCAPTCHA token in it.
* Go to recaptcha console p1573160951123500-slack-CJEQFAAJU and you should see events there (unfortunately I can't see any way to filter the actions)
* Look for track event (instructions in D34725-code)
* Do it again for the passwordless user step
* Test normal account creation and passwordless account creation in the `control` group of the `userStepRecaptcha` A/B test to ensure there are regressions.

andrewserong committed:
Add reCAPTCHA v3 to user step on load, and submit token with new user API call

andrewserong committed:
Update loading reCaptcha script so that it happens in isolation from the other tracking scripts

andrewserong commented:
I've updated this one so that it only loads the reCaptcha script from the `recordgoogleRecaptchaCalypso` function, instead of from `loadTrackingScripts` where it was being loaded on any of the events that other trackers use. The use-case for reCaptcha feels a bit separate to me, and we wouldn't want the reCaptcha badge showing up on unexpected pages.

Some other things to look into:

* [ ] Should we de-load the script after the user step, or leave it running in the background so that it can continue to run all the way up until a site is created? And then, should we intentionally de-load it from the flow controller, or let it automatically unload when the full site reload occurs when a site is created?
* [ ] The code in `ad-tracking.js` is still a little hacky, and we might want to extend it to support other actions than just the hard coded `calypso` action. Could the functions there be generalised a bit more, and perhaps have a method in `analytics/index.js` in the analytics object to call these functions from `ad-tracking.js` instead of importing from `ad-tracking` directly in the user step?

p-jackson committed:
Can record any action name in reCAPTCHA

Refactored recaptcha code to use promises and to accept the action name
as an argument.

p-jackson commented:
I've generalised the code in `ad-tracking.js` so the `recordGoogleRecaptchaAction` function now accepts an action name. Still recording the action on component mount, but I wonder if submitting the form would be a better action to record. It gives recaptcha a few more seconds to analyse the behaviour and maybe having a click even to analyse is more accurate?

W.r.t. de-loading the script, personally I'd like it to disappear asap but haven't found an official API for doing this. We could manually remove the `<script>` tag but will that actually stop the already loaded tracking code from running?

andrewserong commented:
Nice refactor @p-jackson — thanks for looking into this one!

andrewserong committed:
Use explicit render for the reCAPTCHA widget, attached to a div that only exists in the User step

andrewserong committed:
Swap checks at top of recordGoogleRecaptchaAction so that the reCaptcha script doesn't get loaded if isGoogleRecaptchaEnabled is set to false

andrewserong committed:
Pass clientId from grecaptcha.render to grecaptcha.execute

andrewserong commented:
@p-jackson this is slightly hacky, but I've updated the PR to explicitly render the reCAPTCHA badge / widget, and then the DOM element gets removed when the User step is completed. Here's a bit more info in the reCAPTCHA docs: https://developers.google.com/recaptcha/docs/faq#how-can-i-customize-recaptcha-v3 — using `grecaptcha.render` creates the badge, and returns a clientId (starting at 0 it seems to be a unique ID of however many instances of reCaptcha we have running on a page).

I haven't had much luck trying to de-load reCAPTCHA altogether, and from what I've read so far, it sounds like the intention of v3 is that it's kind of always in the background. Deleting `window.grecaptcha` didn't remove everything, and removing the `<script>` tag didn't work for me (I assume since it's all already executed, removing from the DOM doesn't make any difference).

I'll keep at this tomorrow, and look into firing off an additional event when the user submits the User step, but for now, at least the badge is getting removed after the step is submitted...

p-jackson commented:
Nice, that actually puts a few puzzle pieces together I think! I read about a `.reset` method that sounded like it might be able to uninitialise instances. I couldn't figure out how to get the client id though so that could be it!

andrewserong commented:
I ran out of time today to push _clean_ changes, so I haven't pushed anything to this PR just yet. From experimenting with it, it looks like the `.reset` method _resets_ an existing widget, and reloads the widget, rather than removing an instance. And calling `grecaptcha.render` again increments the clientId.

So I feel like me might need to live with the idea that reCaptcha is keeping track of its instance beyond the scope of that first step. Since we run `grecaptcha.render` each time the User step loads, we can know that it'll always be working with a fresh (new) reCaptcha instance.

The do the next bit of this task (adding an additional reCaptcha event when the form is submitted), I think we'll need to do two additional things:

* [ ] Pass the clientId of the reCaptcha instance to the User step to save somewhere
* [ ] Use this clientId in a separate call to `grecaptcha.execute` to fire off the event when the form is submitted, and register the normal form submit function to the callback of that `grecaptcha.execute` so that we have the latest token before submitting the user step.

I could be missing something else, but I'll continue on with this again tomorrow!

p-jackson commented:
Sounds good @andrewserong, I had a play getting the clientId back into the component on this toy branch: `add/recaptcha-action-on-signup-click`

https://github.com/Automattic/wp-calypso/compare/add/recaptcha-v3-to-user-step...add/recaptcha-action-on-signup-click

It's starting to feel a bit messy though. I don't think it's necessarily `ad-tracking` that's the issue, more the way that password/passwordless UI is spread across multiple places.
Don't really have a solution 😅

p-jackson committed:
Add public recaptcha site key to config

p-jackson committed:
Remove recaptcha from e2e tests

Keeps the e2e bots out of the recaptcha dashboard

andrewserong commented:
Thanks for the toy branch @p-jackson and for adding in the site key — it is a bit tricky passing around the clientId, particularly with the way that passwordless is setup. I'll play around with this some more this morning!

andrewserong committed:
Generalise recording a reCaptcha action, and add additional action call before submitting user forms

andrewserong commented:
@p-jackson I've done some further updates in https://github.com/Automattic/wp-calypso/pull/37170/commits/2a59554c27499b9bb4deb3029ff406d4ffce52bf based on the ideas in your branch — I've generalised the `recordGoogleRecaptchaAction` and made init / loading reCaptcha slightly separate. This enables adding in a separate call for another action before submitting the user step (either normal or passwordless). I _think_ the way I've done it means that those steps will only be submitted after reCaptcha returns successfully, so we might need some better error handling here...

What do you think? Does this look too hacky?

To differentiate the event that's fired when the page loads versus when the form's submitted, I've renamed the actions to `calypso/pageload` and `calypso/formSubmit` (the action parameter apparently [only allows alphanumeric characters and slashes](https://developers.google.com/recaptcha/docs/v3)). Again, I'm happy for us to rename these to anything else that suits.

If this direction feels a bit too hacky, I'm happy to revert any of the changes, and feel free to push any of your own changes if you have ideas on how to improve this, too!

p-jackson commented:
I was thinking of using `signup/` as the namespace. `signup/formSubmit` is more specific I think. Could be `calypso/signup/formSubmit` though.

I do like how rendering and sending an action are split into two steps now, passing the clientId down to child components seems fine.

If a human (or bot) submits the form really fast, they could submit before we have a clientId. In that case we'd fail to send the recaptcha action. Couple of ways I thought we could fix that:
- Disable submit button until `initGoogleRecaptcha` has resolved
- Stash the promise from the `initGoogleRecaptcha` call and await it in the `submitForm` method. Actually that'd probably be weird because we'd need to pass a promise object down to the `<PasswordlessSignupForm>` component. Seems like a strange sort of pattern.

So maybe disabling the submit button until `initGoogleRecaptcha` is done? Remembering that even if we get an invalid clientId back we still want the button to get enabled I think. Don't want to block the user because we can't communicate with Google atm.

@andrewserong

p-jackson committed:
Rename signup recaptcha events

Adding `signup` to namespace so it's clearer where the actions come
from.

Switched to `formSubmit` to follow a nounVerb pattern which seems more
common in our track event naming conventions.

p-jackson committed:
Only disable signup button once recaptcha has been initialised

p-jackson commented:
I've updated the PR description and testing instructions. Could you take a look at what I've changed with the submit button @andrewserong? I think this is close to reviewable.

andrewserong committed:
Add disableSubmitButton to PasswordlessSignupForm

andrewserong commented:
Looking great @p-jackson — I've just pushed a tiny extra change to add the disableSubmitButton behaviour to the PasswordlessSignupForm, too. You can't really tell, because that form is disabled anyway if there isn't a valid email address, but I tested it by throttling the network on my browser to slow 3G and managed to quickly get an address in there to test the behaviour, and all looks good.

I'll move this over to ready for review, so we can get some other eyes on it. Thanks again for all your help & collaboration with this one!

andrewserong commented:
And the test instructions are great, thanks 👍

andrewserong committed:
Add userStepRecaptcha A/B test and place reCaptcha behaviour behind abtest checks

andrewserong commented:
Added `userStepRecaptcha` 50/50 A/B test and placed behaviour behind `abtest` checks.

andrewserong committed:
Fix abtest mock in User step test to return an abtest object

p-jackson commented on client/signup/steps/user/index.jsx:
```diff
@@ -200,16 +205,24 @@ export class UserStep extends Component {
+				form: formWithoutPassword,
+				queryArgs: ( this.props.initialContext && this.props.initialContext.query ) || {},
+			} );
+		}
```
@andrewserong To avoid the duplication of logic in the `if` and `else` blocks (also applies to `passwordless.jsx`) what if it was something like this?

```jsx
const recaptchaPromise = 'show' === abtest( 'userStepRecaptcha' ) ? recordGoogleRecaptchaAction( this.state.recaptchaClientId, 'calypso/signup/formSubmit' ) : Promise.resolve();

recaptchaPromise.then( token => {
	this.submit( {
		userData,
		form: formWithoutPassword,
		queryArgs: ( this.props.initialContext && this.props.initialContext.query ) || {},
		recaptchaToken: token || undefined,
	} );
} );
```

`Promise.resolve()` makes a promise that resolves immediately.

andrewserong replied:
Hey, that's a nice one @p-jackson — thanks, I'll give that a shot.

andrewserong replied:
Done! That looks a bit cleaner. Ready for another look @p-jackson

andrewserong committed:
Remove duplicate API calls in user step and passwordless by wrapping recordGoogleRecaptchAction in recaptchaPromise

p-jackson commented on client/blocks/signup-form/passwordless.jsx:
```diff
@@ -76,14 +78,24 @@ class PasswordlessSignupForm extends Component {
-			.then( response => this.createUserAccountFromEmailAddressCallback( null, response ) )
-			.catch( err => this.createUserAccountFromEmailAddressCallback( err ) );
+		const recaptchaPromise =
+			'show' === abtest( 'userStepRecaptcha' )
```
flow check

p-jackson commented on client/signup/steps/user/index.jsx:
```diff
@@ -47,6 +49,8 @@ export class UserStep extends Component {
 		submitting: false,
 		subHeaderText: '',
+		recaptchaClientId: null,
+		isLoadingRecaptcha: 'show' === abtest( 'userStepRecaptcha' ) ? true : false,
```
flow check

p-jackson commented on client/signup/steps/user/index.jsx:
```diff
@@ -74,6 +78,12 @@ export class UserStep extends Component {
 	}
 
 	componentDidMount() {
+		if ( 'show' === abtest( 'userStepRecaptcha' ) ) {
```
flow check

p-jackson commented on client/signup/steps/user/index.jsx:
```diff
@@ -189,10 +205,18 @@ export class UserStep extends Component {
-			form: formWithoutPassword,
-			queryArgs: ( this.props.initialContext && this.props.initialContext.query ) || {},
+		const recaptchaPromise =
+			'show' === abtest( 'userStepRecaptcha' )
```
flow check

p-jackson reviewed (changes requested):
Looks good! The only other thing I think is that we only want to test this "on the default Signup flow". I just added a comment to all those places where we need to add the check for `flow === 'onboarding'`

paObgF-CY-p2

andrewserong committed:
Add flow name check before each abtest check to make sure that we're in the default onboarding flow

andrewserong commented:
Good thinking @-jackson — I've just updated it to add `flowName` checks. I also added it to the one in passwordless, even though that component is currently only available in the onboarding flow, I thought it's better to be explicit about it rather than implicitly relying on how another A/B test is set up.

p-jackson commented:
LGTM! Those tests look like those jetpack tests are failing because they're getting onto the passwordless signup screen, which I'm pretty sure they're not supposed to.

p-jackson committed:
Enable abtest in passwordless test before clicking any buttons

p-jackson commented:
The passwordless e2e test was failing until I switched the order of the steps in d267c472247d8aeb3129bf2ebe9397fdad7af2b4.

I believe what was going wrong was that the `passwordlessSignup` AB test was actually being enabled after navigating to the user step, but because it was originally done by calling a member of `jetpackAddNewSitePage`, it ended up waiting for the wrong DOM element (specifically `.jetpack-new-site__main`).

I'm not sure why this wasn't failing before though, maybe just a timing thing? Pinging @Stojdza in case this is a bad idea for some reason.

Stojdza commented:
Looks good @p-jackson! I was already planning to land this change (77e0017) while I was looking at some flaky test runs. Thanks! 🚀

Stojdza commented:
@p-jackson I went ahead and merged this change in https://github.com/Automattic/wp-calypso/pull/37400. I needed to do this to prevent further failures since _Passwordless_ test was failing against master and live branches.

p-jackson committed:
[reCaptcha v3]: Hide reCAPTCHA badge on phone and show our own… (#37204)

* Hide reCAPTCHA badge on phone and show our own message

Trying out just leaving the badge where it is on desktop. The signup is
already very wordy and at least it's tucked out of the way. However on
small screens the floating badge gets in the way.

Using the same breakpoint that controls the buttons getting larger.

The decision about whether to show the reCAPTCHA message is made by the
signup step component (by passing the `showReCaptchaToS` prop) because
that's where the call to the reCAPTCHA API is initiated.

When hiding the reCAPTCHA the wording is specified by Google at
https://developers.google.com/recaptcha/docs/faq#id-like-to-hide-the-recaptcha-badge.-what-is-allowed

* More padding around recaptcha tos and less text contrast

* Use the correct color-studio code

* Don't adjust padding when user isn't in abtest

* Rename showRecaptchaToS prop

Everywhere else we're treating reCAPTCHA as a single word
